### PR TITLE
Test: trainer integration test and system integration test editing

### LINKED ISSUE #644: [TEST] Trainer Integration
### What do you want to test?
Full integration test for trainer.

### Outline of test structure
Should use all components typically found in the trainer and run without error, but not using launchpad. See the jax IPPO example for how the components are typically logically distributed between core nodes. If multiple tests are required to cover different distinct cases (e.g. value-based vs policy-gradient), then they should be included.

### Definition of done
Test(s) passing and including all typical components.

### Mandatory checklist before making a PR
* [ ] The success criteria laid down in “Definition of done” are met.
* [ ] Test code is documented - docstrings for methods and classes, static types for arguments.
* [ ] Documentation is updated - README, CONTRIBUTING, or other documentation.

### PR DESCRIPTION
**Pull request by OmaymaMahjoub**
## What?
Trainer integration test
Edit the system integration test to work in the multi-process case #542 

## Why?
Part of on going efforts to increase testing code coverage of callback redesigned Mava.

## Extra
Close #644

### CODE REVIEW

File: tests/jax/integration/trainer_test.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
+import functools
```
Comment (AsadJeewa) [inline tests/jax/integration/trainer_test.py:1]: please add the license at the top of this file

Comment (OmaymaMahjoub) [reply tests/jax/integration/trainer_test.py:18]: Done!

File: tests/jax/components/building/distributor_test.py [hunk 2]
```diff
@@ -119,27 +119,34 @@ def test_on_building_program_nodes_multi_process(
         "evaluator",
         "trainer",
     ]
+
```
Comment (AsadJeewa) [inline tests/jax/components/building/distributor_test.py:122]: see above (line 50 and elsewhere). Think the docstring should change from "to test no multi_process" to "to test single-process"

File: tests/jax/components/building/distributor_test.py [hunk 3]
```diff
@@ -162,23 +169,28 @@ def test_on_building_program_nodes_multi_process_no_evaluator(
...
-    assert (
-        mock_builder.store.program._program._groups["parameter_server"][-1]._constructor
-        == mock_builder.parameter_server
+        ]._constructor()
+        == "Parameter Server Test"
     )
     assert (
-        mock_builder.store.program._program._groups["executor"][-1]._constructor
-        == mock_builder.executor
+        mock_builder.store.program._program._groups["executor"][-1]._constructor(
```
Comment (AsadJeewa) [inline tests/jax/components/building/distributor_test.py:184]: Think a comment about why we need to index -1 would be useful

File: tests/jax/integration/trainer_test.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+    (
+        data_server,
+        parameter_server,
+        executor,
+        evaluator,
+        trainer,
+    ) = test_system._builder.store.system_build
+
+    for i in range(0, 4):
```
Comment (AsadJeewa) [inline tests/jax/integration/trainer_test.py:113]: 0 is reduendant so can be removed. Should just run for `for i in range(5)` like the mappo test

File: tests/jax/integration/trainer_test.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    )
+
+    # Disable the run of the trainer node
+    (trainer_node,) = test_system._builder.store.program._program._groups["trainer"]
+    trainer_node.disable_run()
+    test_address_builder.bind_addresses([trainer_node])
+
+    # launch the system
+    test_system.launch()
+    time.sleep(10)  # wait the executor to run at least one time
```
Comment (AsadJeewa) [inline tests/jax/integration/trainer_test.py:199]: "wait till the executor has run"

File: tests/jax/integration/trainer_test.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        sample_batch_size=5,
+        checkpoint=True,
+        nodes_on_gpu=[],
+        lp_launch_type=lp.LaunchType.TEST_MULTI_THREADING,
+    )
+
+    # Disable the run of the trainer node
+    (trainer_node,) = test_system._builder.store.program._program._groups["trainer"]
+    trainer_node.disable_run()
+    test_address_builder.bind_addresses([trainer_node])
```
Comment (AsadJeewa) [inline tests/jax/integration/trainer_test.py:195]: the test still seems to pass when I remove this line. Same for removing the sleep (though I get the need for sleep)

File: tests/jax/integration/trainer_test.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            assert jnp.all(categorical_value_head["w"] == 0)
+
+    # Step function
+    trainer.step()
+
+    # Check that the trainer update the network
+    for net_key in trainer.store.networks["networks"].keys():
+        mu = trainer.store.opt_states[net_key][1][0][-1]
+        for categorical_value_head in mu.values():
+            assert not jnp.all(categorical_value_head["b"] == 0)
```
Comment (AsadJeewa) [inline tests/jax/integration/trainer_test.py:217]: really like this check :smile:

**Review by AsadJeewa — CHANGES_REQUESTED** [review]
Thanks @OmaymaMahjoub 
Really nice work on this. Few small changes requested

File: tests/jax/integration/trainer_test.py (hunk 6 above)
Comment (OmaymaMahjoub) [reply tests/jax/integration/trainer_test.py:195]: I think we need to keep the time.sleep because it depends on the speed of the device that you have! For the bind it is needed else from my side it crushes  and it was mentioned in the tests in Launchpad

Comment (OmaymaMahjoub) [reply tests/jax/integration/trainer_test.py:193]: You were right we don't need the `test_address_builder.bind_addresses([trainer_node])` for the test to pass

File: tests/jax/components/building/distributor_test.py (hunk 3 above)
Comment (AsadJeewa) [reply tests/jax/components/building/distributor_test.py]: discussed offline

File: tests/jax/integration/trainer_test.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+    # After run step method
+    for net_key in trainer.store.networks["networks"].keys():
+        mu = trainer.store.opt_states[net_key][1][0][-1]
+        for categorical_value_head in mu.values():
+            assert not jnp.all(categorical_value_head["b"] == 0)
+            assert not jnp.all(categorical_value_head["w"] == 0)
+
+
+def test_trainer_multi_thread(test_system: System) -> None:
```
Comment (AsadJeewa) [inline tests/jax/integration/trainer_test.py:129]: since our variable is multi_process, I think that this should be called `test_trainer_multi_process()` still

**Review by AsadJeewa — CHANGES_REQUESTED** [review]
Small request (which is not essential), then happy to approve. Thanks @OmaymaMahjoub

File: tests/jax/integration/trainer_test.py (hunk 8 above)
Comment (OmaymaMahjoub) [reply tests/jax/integration/trainer_test.py:129]: But in this case aren't we having lp_launch_type=lp.LaunchType.TEST_MULTI_THREADING and probably in the future we will add the test for lp_launch_type=lp.LaunchType.TEST_MULTI_PROCESSING so I thought to keep it as that properly!
What do you think?

**Review by DriesSmit — DISMISSED** [review]
This looks great, thanks @OmaymaMahjoub and @RuanJohn :fire: Really glad we can test multi_process now.

File: tests/jax/integration/trainer_test.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+    # Check that the trainer update the network
+    for net_key in trainer.store.networks["networks"].keys():
+        mu = trainer.store.opt_states[net_key][1][0][-1]
+        for categorical_value_head in mu.values():
+            assert not jnp.all(categorical_value_head["b"] == 0)
+            assert not jnp.all(categorical_value_head["w"] == 0)
+
+
+def test_trainer_multi_process(test_system_mp: System) -> None:
```
Comment (AsadJeewa) [inline tests/jax/integration/trainer_test.py:108]: The only issue is that the test (passes) does not terminate

**Review by AsadJeewa — COMMENTED** [review]
Lets discuss the termination of the test offline

File: tests/jax/integration/trainer_test.py (hunk 9 above)
Comment (OmaymaMahjoub) [reply tests/jax/integration/trainer_test.py:108]: It won't make any problem since it is a test that's what I understood, also the checks passes them so it won't bother
what do you think?

Comment (AsadJeewa) [reply tests/jax/integration/trainer_test.py:108]: Ideally, we should use the terminators to fix this

Comment (OmaymaMahjoub) [reply tests/jax/integration/trainer_test.py:108]: It won't work unfortunately we need to come up with another idea

**Review by AsadJeewa — APPROVED** [review]
Thanks @OmaymaMahjoub

**Review by AsadJeewa — APPROVED** [review]

**Review by DriesSmit — APPROVED** [review]
Thanks @OmaymaMahjoub :slightly_smiling_face:

File: tests/jax/integration/trainer_test.py (hunk 9 above)
Comment (OmaymaMahjoub) [reply tests/jax/integration/trainer_test.py:111]: We solved that for the time being using `os.kill()`, however I will create an issue to add a feature that give us the possibility of using `lp.stop()` within the Terminator components
